build,pbsync: various tweaks to make the tool run on Linux

commodo opened the pull request:
Hey,

While trying to figure out how to build the game from source, I tried to use the PBSync tool to see what it does.
After I while I think I figure out partially how to use it [without any docs].

Trying to run this on a Linux machine, this is where I got to, before hitting a dead-end.
Maybe will continue on Discord to get any hints [if possible].

Note: the `Signed-off-by` tag is a habit; most open-source projects require it as a form of legalese; I can remove it if necessary.

Signed-off-by: Alexandru Ardelean <ardeleanalex@gmail.com>

mastercoms commented:
Thanks for the catch about sequence vs. strings for shell execution. We had to enable shell on Windows or else git & friends wouldn't be found on path. So we just enabled it without seeing the other side effects to other args.

As for the GCM stuff, that's required. I just hadn't updated the Base-Project PBSync.xml yet, but I have now, for your convenience. Usually we do not include backwards compatibility, and expect old projects to upgrade to accommodate changes in PBSync.

Finally, I don't mind the sign-off tag.

commodo commented:
> As for the GCM stuff, that's required. I just hadn't updated the Base-Project PBSync.xml yet, but I have now, for your convenience. Usually we do not include backwards compatibility, and expect old projects to upgrade to accommodate changes in PBSync.

ack
will drop that commit
thanks for the quick fix;
maybe, i'll change it to checking which attribute is missing and report that;
getting a 'NoneType' has no 'text' attribute error requires some debugging;

commodo commented:
> Thanks for the catch about sequence vs. strings for shell execution. We had to enable shell on Windows or else git & friends wouldn't be found on path. So we just enabled it without seeing the other side effects to other args.

I think this is very dependent on the environment that you use [normal Windows, MinGW, MSYS, etc?].
Technically if you ensure that the PATH env var is fine git & friends should be found-able.
But I do remember, that on Windows, Git is usually (or used to be) deployed packaged with some sort of MinGW/MSYS environment which has some Linux-like bash on top of Windows.
So, it becomes a bit quirky.

commodo commented:
on a side note;
i'm also waiting for this to get through:   https://github.com/GoogleCloudPlatform/gsutil/pull/1174
after that the note about manually patching gsutil should go away   ->  simpler life

commodo commented:
Changelog v1 -> v2:
* dropped commit `pbsync/pbsync.py: handle git credential managers configs as optional `
* added commit `pbsync/pbsync.py: rework config parser func to raise informative exception` ; this should highlight better which config options are missing [if any]

mastercoms commented on pbsync/pbsync.py:
```diff
@@ -413,27 +413,46 @@ def main(argv):
+        missing_keys = []
+        config_map = {}
+        for key, val in config_args_map.items():
+            key1, override = val
```
I would see more value in this being a fallback/default, rather than an override. If it uses a fallback, then the missing key error can just be logged rather than raising an exception.

commodo replied:
ack
the reason this ended up being like this, is because it was a bit closer to the original behavior;

but i agree that it can be inverted a bit and have and make it have a default/fallback behavior vs override;

commodo replied:
thought about it a bit more;
so, just to make sure we're on the same page;

the old behavior for `expected_branch_name` was: 
```
'expected_branch_name': root.find('git/expectedbranch').text if args.debugbranch is None else str(args.debugbranch),
```

i.e. (in pseudo-code)
```
if args.debugbranch is None:
    'expected_branch_name' = root.find('git/expectedbranch').text
else
    'expected_branch_name' = str(args.debugbranch)
```

if you reverse the condition, it looks more like an override, than a default; 
because `args.debugbranch` is preferred if provided;
so [reversing things a bit]:
```
if args.debugbranch [ is not  None]:
    'expected_branch_name' = str(args.debugbranch)
else
    'expected_branch_name' = root.find('git/expectedbranch').text
```

but i am fine to change the default logic;
but i wanted to highlight this subtlety;

mastercoms replied:
Ah, I didn't notice your usage for the debug branch arg. Yeah the current implementation seems fine then.
